File: backend/app/services/user_service.py
from typing import List, Optional, Dict, Any
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_, func
from ..models.user import User, UserRole
from ..schemas.user import UserCreate, UserUpdate, UserResponse, UserList, UserListResponse, UserPasswordChange, UserBulkAction, UserCreateByAdmin
from ..core.auth import get_password_hash, verify_password

class UserService:
    def __init__(self, db: Session):
        self.db = db

    def create_user(self, user_data: UserCreate) -> UserResponse:
        """Cria um novo usuário"""
        # Verificar se email já existe
        existing_user = self.db.query(User).filter(User.email == user_data.email).first()
        if existing_user:
            raise ValueError("Email já está em uso")

        # Hash da senha
        user_dict = user_data.model_dump()
        user_dict['password_hash'] = get_password_hash(user_dict.pop('password'))

        # Garantir que novos usuários começam como Visualizador, a menos que explicitamente definido
        user_dict.setdefault('role', UserRole.VIEWER)

        # Criar usuário
        user = User(**user_dict)
        self.db.add(user)
        self.db.commit()
        self.db.refresh(user)

        return UserResponse.model_validate(user)

    def create_user_as_admin(self, user_data: 'UserCreateByAdmin') -> UserResponse:
        """
        Cria usuário pelo administrador - cria tanto no Supabase Auth quanto na tabela users
        Permite que o usuário faça login imediatamente após definir senha
        """
        from config.supabase import get_supabase_admin
        import secrets

        # Verificar se email já existe na nossa tabela
        existing_user = self.db.query(User).filter(User.email == user_data.email).first()
        if existing_user:
            raise ValueError("Email já está em uso")

        try:
            # 1. Criar usuário no Supabase Auth usando service role key
            supabase_admin = get_supabase_admin()

            # Gerar senha temporária aleatória (usuário mudará via email)
            temp_password = secrets.token_urlsafe(16)

            auth_response = supabase_admin.auth.admin.create_user({
                "email": user_data.email,
                "password": temp_password,
                "email_confirm": False,  # Usuário precisa confirmar email
                "user_metadata": {
                    "full_name": user_data.full_name,
                    "role": user_data.role.value
                }
            })

            if not auth_response or not auth_response.user:
                raise ValueError("Falha ao criar usuário no Supabase Auth")

            auth_user_id = auth_response.user.id

            # 2. Enviar email de redefinição de senha (usuário define própria senha)
            if user_data.send_welcome_email:
                supabase_admin.auth.admin.generate_link({
                    "type": "recovery",
                    "email": user_data.email
                })

            # 3. Criar registro na tabela users
            user = User(
                auth_id=str(auth_user_id),
                email=user_data.email,
                full_name=user_data.full_name,
                role=user_data.role,
                phone=user_data.phone,
                bio=user_data.bio,
                permissions_json=user_data.permissions_json or {},
                is_active=True,
                password_hash=get_password_hash(temp_password)  # Fallback, não será usado
            )

            self.db.add(user)
            self.db.commit()
            self.db.refresh(user)

            return UserResponse.model_validate(user)

        except Exception as e:
            self.db.rollback()
            # Se falhou, tentar limpar usuário do Auth se foi criado
            raise ValueError(f"Erro ao criar usuário: {str(e)}")

    def get_user(self, user_id: int) -> Optional[UserResponse]:
        """Obtém um usuário por ID"""
        user = self.db.query(User).filter(User.id == user_id).first()
        if not user:
            return None

        return UserResponse.model_validate(user)

    def get_users(self, skip: int = 0, limit: int = 100, search: str = None, role: UserRole = None, is_active: bool = None) -> UserListResponse:
        """Lista usuários com filtros e paginação"""
        query = self.db.query(User)

        if search:
            query = query.filter(
                or_(
                    User.full_name.ilike(f"%{search}%"),
                    User.email.ilike(f"%{search}%"),
                    User.phone.ilike(f"%{search}%")
                )
            )

        if role:
            query = query.filter(User.role == role)

        if is_active is not None:
            query = query.filter(User.is_active == is_active)

        # Contar total
        total = query.count()

        # Aplicar paginação
        users = query.offset(skip).limit(limit).all()

        # Calcular total de páginas
        total_pages = (total + limit - 1) // limit

        return UserListResponse(
            users=[UserList.model_validate(user) for user in users],
            total=total,
            page=skip // limit + 1,
            size=limit,
            total_pages=total_pages
        )

    def update_user(self, user_id: int, user_data: UserUpdate) -> Optional[UserResponse]:
        """Atualiza um usuário"""
        user = self.db.query(User).filter(User.id == user_id).first()
        if not user:
            return None

        update_data = user_data.model_dump(exclude_unset=True)

        # Verificar se email já existe (se estiver sendo alterado)
        if 'email' in update_data and update_data['email'] != user.email:
            existing_user = self.db.query(User).filter(User.email == update_data['email']).first()
            if existing_user:
                raise ValueError("Email já está em uso")

        # Hash da senha se fornecida
        if 'password' in update_data:
            update_data['password_hash'] = get_password_hash(update_data.pop('password'))

        for field, value in update_data.items():
            setattr(user, field, value)

        self.db.commit()
        self.db.refresh(user)

        return UserResponse.model_validate(user)

    def update_user_permissions(self, user_id: int, permissions: Dict[str, Any]) -> Optional[UserResponse]:
        """Atualiza permissões personalizadas do usuário."""
        user = self.db.query(User).filter(User.id == user_id).first()
        if not user:
            return None

        # Sanitizar para apenas valores booleanos (rotas específicas)
        sanitized = {
            key: bool(value) for key, value in (permissions or {}).items()
        }

        user.permissions_json = sanitized
        self.db.commit()
        self.db.refresh(user)

        return UserResponse.model_validate(user)

    def delete_user(self, user_id: int) -> bool:
        """Remove um usuário (soft delete)"""
        user = self.db.query(User).filter(User.id == user_id).first()
        if not user:
            return False

        # Soft delete - marcar como inativo
        user.is_active = False
        self.db.commit()
        return True

    def activate_user(self, user_id: int) -> bool:
        """Ativa um usuário"""
        user = self.db.query(User).filter(User.id == user_id).first()
        if not user:
            return False

        user.is_active = True
        self.db.commit()
        return True

    def change_password(self, user_id: int, password_data: UserPasswordChange) -> bool:
        """Altera senha do usuário"""
        user = self.db.query(User).filter(User.id == user_id).first()
        if not user:
            return False

        # Verificar senha atual
        if not verify_password(password_data.current_password, user.password_hash):
            raise ValueError("Senha atual incorreta")

        # Atualizar senha
        user.password_hash = get_password_hash(password_data.new_password)
        self.db.commit()
        return True

    def bulk_action(self, action_data: UserBulkAction) -> Dict[str, Any]:
        """Executa ação em lote nos usuários"""
        results = {"success": [], "errors": []}

        for user_id in action_data.user_ids:
            try:
                user = self.db.query(User).filter(User.id == user_id).first()
                if not user:
                    results["errors"].append({"id": user_id, "error": "Usuário não encontrado"})
                    continue

                if action_data.action == "activate":
                    user.is_active = True
                elif action_data.action == "deactivate":
                    user.is_active = False
                elif action_data.action == "delete":
                    user.is_active = False  # Soft delete
                elif action_data.action == "change_role" and action_data.role:
                    user.role = action_data.role
                else:
                    results["errors"].append({"id": user_id, "error": "Ação inválida"})
                    continue

                self.db.commit()
                results["success"].append({"id": user_id, "email": user.email})

            except Exception as e:
                results["errors"].append({"id": user_id, "error": str(e)})

        return results

    def get_users_by_role(self, role: UserRole) -> List[UserList]:
        """Obtém usuários por role"""
        users = self.db.query(User).filter(User.role == role, User.is_active == True).all()
        return [UserList.model_validate(user) for user in users]

    def get_user_stats(self) -> Dict[str, Any]:
        """Obtém estatísticas de usuários"""
        total_users = self.db.query(User).count()
        active_users = self.db.query(User).filter(User.is_active == True).count()

        role_stats = {}
        for role in UserRole:
            count = self.db.query(User).filter(User.role == role, User.is_active == True).count()
            role_stats[role.value] = count

        return {
            "total_users": total_users,
            "active_users": active_users,
            "inactive_users": total_users - active_users,
            "role_distribution": role_stats
        }

    def get_users_for_assignment(self) -> List[Dict[str, Any]]:
        """Obtém usuários disponíveis para atribuição"""
        users = self.db.query(User).filter(User.is_active == True).all()
        return [
            {
                "id": user.id,
                "full_name": user.full_name,
                "email": user.email,
                "role": user.role.value,
                "phone": user.phone
            }
            for user in users
        ]

    def get_user_activity_summary(self, user_id: int) -> Dict[str, Any]:
        """Obtém resumo de atividades do usuário"""
        user = self.db.query(User).filter(User.id == user_id).first()
        if not user:
            return {}

        return {
            "user_id": user_id,
            "full_name": user.full_name,
            "email": user.email,
            "role": user.role.value,
            "is_active": user.is_active,
            "created_at": user.created_at,
            "updated_at": user.updated_at
        }

File: frontend/src/services/userService.ts
// src/services/userService.ts
import { supabase } from '../config/supabaseClient';
import {
  User,
  UserRole,
  UserList,
  UserCreate,
  UserUpdate,
  UserPasswordChange,
  UserBulkAction,
  UserStats,
} from '../types/user';
import { withDiagnostic } from '../utils/withDiagnostic';

// Cache de requests em andamento para evitar duplicatas
const pendingRequests = new Map<string, Promise<any>>();

/**
 * Wrapper para requests com deduplication
 * Evita múltiplas chamadas simultâneas para o mesmo recurso
 */
async function withDeduplication<T>(
  key: string,
  requestFn: () => Promise<T>
): Promise<T> {
  // Se já existe uma request em andamento, reutiliza
  if (pendingRequests.has(key)) {
    console.log(`🔄 Reusing pending user request for: ${key}`);
    return pendingRequests.get(key) as Promise<T>;
  }

  // Cria nova request
  const request = requestFn().finally(() => {
    // Remove do cache quando completar
    pendingRequests.delete(key);
  });

  // Armazena no cache
  pendingRequests.set(key, request);

  return request;
}

export const userService = {
  // Create user
  createUser: (userData: UserCreate) =>
    withDiagnostic('userService.createUser', async () => {
      const payload = { ...userData, password_hash: 'managed_by_auth' };
      // @ts-ignore: remove password if present
      delete payload.password;
      const { data, error } = await supabase
        .from('users')
        .insert([payload])
        .select()
        .single();
      if (error) throw error;
      return data as User;
    }),

  // Create user as admin (with Supabase Auth integration)
  createUserAsAdmin: (userData: {
    email: string;
    full_name: string;
    role: UserRole;
    permissions_json?: Record<string, boolean>;
    phone?: string;
    bio?: string;
  }) =>
    withDiagnostic('userService.createUserAsAdmin', async () => {
      const { data, error } = await supabase
        .from('users')
        .insert([
          {
            email: userData.email,
            full_name: userData.full_name,
            role: userData.role,
            permissions_json: userData.permissions_json || {},
            phone: userData.phone,
            bio: userData.bio,
            is_active: true,
            password_hash: '', // Managed by Supabase Auth
          },
        ])
        .select()
        .single();
      if (error) throw error;
      return data as User;
    }),

  // Get users with filters & pagination
  getUsers: (params?: {
    skip?: number;
    limit?: number;
    search?: string;
    role?: UserRole;
    is_active?: boolean;
  }) => {
    const cacheKey = `users-${JSON.stringify(params || {})}`;

    return withDeduplication(cacheKey, () =>
      withDiagnostic('userService.getUsers', async () => {
        let query = supabase.from('users').select('*', { count: 'exact' });
        if (params?.role) query = query.eq('role', params.role);
        if (params?.is_active !== undefined)
          query = query.eq('is_active', params.is_active);
        if (params?.search) {
          query = query.or(
            `email.ilike.%${params.search}%,full_name.ilike.%${params.search}%`
          );
        }
        if (params?.limit) {
          const from = params.skip || 0;
          const to = from + params.limit - 1;
          query = query.range(from, to);
        }
        const { data, error, count } = await query;
        if (error) throw error;
        const size = params?.limit || 10;
        const page =
          params?.skip && params?.limit
            ? Math.floor(params.skip / params.limit)
            : 0;
        const total = count || 0;
        const total_pages = Math.ceil(total / size);
        return {
          users: (data as UserList[]) || [],
          total,
          page,
          size,
          total_pages,
        };
      })
    );
  },

  // Get single user by ID
  getUser: (userId: number) =>
    withDiagnostic('userService.getUser', async () => {
      const { data, error } = await supabase
        .from('users')
        .select('*')
        .eq('id', userId)
        .single();
      if (error) throw error;
      if (!data) throw new Error('User not found');
      return data as User;
    }),

  // Update user
  updateUser: (userId: number, userData: UserUpdate) =>
    withDiagnostic('userService.updateUser', async () => {
      const { data, error } = await supabase
        .from('users')
        .update(userData)
        .eq('id', userId)
        .select()
        .single();
      if (error) throw error;
      return data as User;
    }),

  // Delete user
  deleteUser: (userId: number) =>
    withDiagnostic('userService.deleteUser', async () => {
      const { error } = await supabase.from('users').delete().eq('id', userId);
      if (error) throw error;
    }),

  // Activate user
  activateUser: (userId: number) =>
    withDiagnostic('userService.activateUser', async () => {
      const { error } = await supabase
        .from('users')
        .update({ is_active: true })
        .eq('id', userId);
      if (error) throw error;
    }),

  // Deactivate user
  deactivateUser: (userId: number) =>
    withDiagnostic('userService.deactivateUser', async () => {
      const { error } = await supabase
        .from('users')
        .update({ is_active: false })
        .eq('id', userId);
      if (error) throw error;
    }),

  // Change password – delegated to auth service
  changePassword: async () => {
    throw new Error('Use authService.changePassword() for password changes');
  },

  // Bulk actions
  bulkAction: (actionData: UserBulkAction) =>
    withDiagnostic('userService.bulkAction', async () => {
      if (actionData.action === 'delete') {
        const { error } = await supabase
          .from('users')
          .delete()
          .in('id', actionData.user_ids);
        if (error) throw error;
      } else if (actionData.action === 'activate') {
        const { error } = await supabase
          .from('users')
          .update({ is_active: true })
          .in('id', actionData.user_ids);
        if (error) throw error;
      } else if (actionData.action === 'deactivate') {
        const { error } = await supabase
          .from('users')
          .update({ is_active: false })
          .in('id', actionData.user_ids);
        if (error) throw error;
      } else if (actionData.action === 'change_role' && actionData.role) {
        const { error } = await supabase
          .from('users')
          .update({ role: actionData.role })
          .in('id', actionData.user_ids);
        if (error) throw error;
      }
      return { success: true, message: 'Bulk action completed' };
    }),

  // Users by role
  getUsersByRole: (role: UserRole) =>
    withDiagnostic('userService.getUsersByRole', async () => {
      const { data, error } = await supabase
        .from('users')
        .select('*')
        .eq('role', role);
      if (error) throw error;
      return (data as UserList[]) || [];
    }),

  // User statistics
  getUserStats: () =>
    withDiagnostic('userService.getUserStats', async () => {
      const [all, active, inactive, roles] = await Promise.all([
        supabase.from('users').select('*', { count: 'exact', head: true }),
        supabase
          .from('users')
          .select('*', { count: 'exact', head: true })
          .eq('is_active', true),
        supabase
          .from('users')
          .select('*', { count: 'exact', head: true })
          .eq('is_active', false),
        supabase.from('users').select('role'),
      ]);
      const roleDist: Record<string, number> = {};
      (roles.data || []).forEach((u: any) => {
        const r = u.role || 'unknown';
        roleDist[r] = (roleDist[r] || 0) + 1;
      });
      return {
        total_users: all.count || 0,
        active_users: active.count || 0,
        inactive_users: inactive.count || 0,
        role_distribution: roleDist,
      };
    }),

  // Users for assignment (active only)
  getUsersForAssignment: () =>
    withDiagnostic('userService.getUsersForAssignment', async () => {
      const { data, error } = await supabase
        .from('users')
        .select('*')
        .eq('is_active', true);
      if (error) throw error;
      return (data as UserList[]) || [];
    }),

  // Current user profile
  getMyProfile: () =>
    withDiagnostic('userService.getMyProfile', async () => {
      const {
        data: { user },
      } = await supabase.auth.getUser();
      if (!user?.email) throw new Error('Not authenticated');
      const { data, error } = await supabase
        .from('users')
        .select('*')
        .eq('email', user.email)
        .single();
      if (error) throw error;
      return data as User;
    }),

  // Update current user profile
  updateMyProfile: (userData: UserUpdate) =>
    withDiagnostic('userService.updateMyProfile', async () => {
      const {
        data: { user },
      } = await supabase.auth.getUser();
      if (!user?.email) throw new Error('Not authenticated');
      const { data, error } = await supabase
        .from('users')
        .update(userData)
        .eq('email', user.email)
        .select()
        .single();
      if (error) throw error;
      return data as User;
    }),

  // Update user permissions
  updateUserPermissions: (
    userId: number,
    permissions: Record<string, unknown>
  ) =>
    withDiagnostic('userService.updateUserPermissions', async () => {
      await userService.updateUser(userId, {
        permissions_json: permissions,
      } as any);
    }),
};

export const getUserRoleLabel = (role: UserRole): string => {
  const labels = {
    [UserRole.ADMIN]: 'Administrador',
    [UserRole.MANAGER]: 'Gerente',
    [UserRole.COORDINATOR]: 'Coordenador',
    [UserRole.OPERATOR]: 'Operador',
    [UserRole.VIEWER]: 'Visualizador',
    [UserRole.CLIENT]: 'Cliente',
    [UserRole.CONTRIBUTOR]: 'Contribuidor (Novo)',
  };
  return labels[role] || role;
};

export const getUserRoleColor = (role: UserRole): string => {
  const colors: Record<string, string> = {
    [UserRole.ADMIN]: 'error',
    [UserRole.MANAGER]: 'warning',
    [UserRole.COORDINATOR]: 'info',
    [UserRole.OPERATOR]: 'primary',
    [UserRole.VIEWER]: 'secondary',
    [UserRole.CLIENT]: 'success',
    [UserRole.CONTRIBUTOR]: 'default',
  };
  return colors[role] || 'default';
};

export default userService;
